### PR TITLE
SE-10166: Added staging mode to Setup.Mode

### PR DESCRIPTION
**Pull request by marvinwoop**
Additionally I've combined the configuration merging methods like applyDeveloperConfig and loadInstanceConfig to a applyConfig, which accepts a String parameter to look for the corresponding files or property names.

### CODE REVIEW

**Review by qw3ry — APPROVED** [review]

File: src/main/java/sirius/kernel/Setup.java [hunk 1]
```diff
@@ -58,7 +58,26 @@
      * Determines the mode in which the framework should run. This mainly effects logging and the configuration.
      */
     public enum Mode {
-        DEV, TEST, PROD
+
+        /**
+         * Defines the IDE development {@link Mode}
```
Comment (andyHa) [inline src/main/java/sirius/kernel/Setup.java:63]: why link to the enum itself. rather use one sentence per value what the environment typically is - esp. staging vs. test etc.

File: src/main/java/sirius/kernel/Sirius.java [hunk 2]
```diff
@@ -287,10 +299,12 @@ private static void setupLocalConfig() {
             // instance.conf and develop.conf are not used to tests to permit uniform behaviour on local
             // machines and build servers...
             if (Sirius.isDev()) {
-                config = setup.applyDeveloperConfig(config);
+                config = setup.applyConfig(config, CONFIG_DEVELOP);
+            } else if (Sirius.isStaging()) {
+                config = setup.applyConfig(config, CONFIG_STAGING);
```
Comment (andyHa) [inline src/main/java/sirius/kernel/Sirius.java:304]: didn't we also want to add a prod.conf? to configure production only settings?

**Review by andyHa — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/kernel/Sirius.java (hunk 2 above)
Comment (marvinwoop) [reply src/main/java/sirius/kernel/Sirius.java:304]: Isn't that effectively our instance.conf?

File: src/main/java/sirius/kernel/Setup.java (hunk 1 above)
Comment (marvinwoop) [reply src/main/java/sirius/kernel/Setup.java]: Thats right, changed it.

File: src/main/java/sirius/kernel/Sirius.java (hunk 2 above)
Comment (marvinwoop) [reply src/main/java/sirius/kernel/Sirius.java:305]: I got it wrong, first. I now have added a productive.conf check.

File: src/main/java/sirius/kernel/Sirius.java [hunk 3]
```diff
@@ -90,14 +94,23 @@ private Sirius() {
...
     public static boolean isDev() {
         return setup.getMode() == Setup.Mode.DEV;
     }
 
+    /**
+     * Determines if the framework is running in staging mode.
+     *
+     * @return {@code true} if the framework runs in staging mode, {@code false} otherwise
+     */
+    public static boolean isStaging() {
```
Comment (qw3ry) [inline src/main/java/sirius/kernel/Sirius.java:110]: `Sirius.isTest()`? Oder brauchen wir das nicht?

Comment (marvinwoop) [reply src/main/java/sirius/kernel/Sirius.java:110]: Evtl. könnte man das mal brauchen aber wirklich brauchen tun wir's aktuell nicht, denke ich. Was meinen die anderen dazu?

File: src/main/java/sirius/kernel/Setup.java [hunk 4]
```diff
@@ -360,48 +379,28 @@ public Config applyTestScenarioConfig(@Nullable String scenarioFile, @Nonnull Co
     }
 
     /**
-     * Applies the developer configuration to the given config object.
+     * Applies the corresponding configuration to the given config object.
      * <p>
-     * By default this loads and applies "develop.conf" from the file system.
+     * By default this loads and applies the given "instance.conf", "develop.conf", "staging.conf" or "productive.cong"
```
Comment (mkeckmkeck) [inline src/main/java/sirius/kernel/Setup.java:384]:
```suggestion
     * By default this loads and applies the given "instance.conf", "develop.conf", "staging.conf" or "productive.conf"
```

File: src/main/java/sirius/kernel/Sirius.java [hunk 5]
```diff
@@ -134,7 +156,7 @@ public static boolean isRunning() {
      * loggers
      */
     private static void setupLogLevels() {
-        if (Sirius.isDev()) {
+        if (!Sirius.isProd()) {
             Log.setLevel(DEBUG_LOGGER_NAME, Level.FINEST);
```
Comment (mkeckmkeck) [inline src/main/java/sirius/kernel/Sirius.java:160]: Javadoc am DEBUG logger noch anpassen

File: src/main/java/sirius/kernel/Sirius.java [hunk 6]
```diff
@@ -287,10 +309,14 @@ private static void setupLocalConfig() {
             // instance.conf and develop.conf are not used to tests to permit uniform behaviour on local
             // machines and build servers...
             if (Sirius.isDev()) {
-                config = setup.applyDeveloperConfig(config);
+                config = setup.applyConfig(config, CONFIG_DEVELOP);
+            } else if (Sirius.isStaging()) {
+                config = setup.applyConfig(config, CONFIG_STAGING);
+            } else if (Sirius.isProd()) {
+                config = setup.applyConfig(config, CONFIG_PRODUCTIVE);
             }
```
Comment (mkeckmkeck) [inline src/main/java/sirius/kernel/Sirius.java:317]: Falls wir Mode.DEV nach Mode.DEVELOP ändern, könnte man die ifs und das mapping sparen und direkt getMode.toString().toLowerCase() an applyConfig übergeben

**Review by mkeckmkeck — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/kernel/Setup.java (hunk 4 above)
Comment (marvinwoop) [reply src/main/java/sirius/kernel/Setup.java]: 👍

File: src/main/java/sirius/kernel/Sirius.java (hunk 6 above)
Comment (marvinwoop) [reply src/main/java/sirius/kernel/Sirius.java]: Gute Idee, danke. Ist umgesetzt.

**Review by mkeckmkeck — APPROVED** [review]

File: src/main/java/sirius/kernel/Setup.java [hunk 7]
```diff
@@ -58,7 +58,26 @@
...
+
+        /**
+         * Defines the staging mode
+         */
+        STAGING,
+
+        /**
+         * Defines the productive mode
+         */
+        PROD
```
Comment (sabieber) [inline src/main/java/sirius/kernel/Setup.java:80]:
```suggestion
        /**
         * Defines the IDE development mode.
         */
        DEVELOP,

        /**
         * Defines the test mode, commonly used by JUnit.
         */
        TEST,

        /**
         * Defines the staging mode.
         */
        STAGING,

        /**
         * Defines the productive mode.
         */
        PROD
```

File: src/main/java/sirius/kernel/Sirius.java [hunk 8]
```diff
@@ -67,9 +67,10 @@
     protected static final Log LOG = Log.get("sirius");
 
     private static final String DEBUG_LOGGER_NAME = "debug";
+    private static final String CONFIG_INSTANCE = "instance";
 
     /**
-     * This debug logger will be logging all messages when {@link sirius.kernel.Sirius#isDev()} is true. Otherwise,
+     * This debug logger will be logging all messages when {@link Sirius#isProd()} is false. Otherwise,
```
Comment (sabieber) [inline src/main/java/sirius/kernel/Sirius.java:73]: Doesnt this a lot more is logged in test/staging?

**Review by sabieber — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/kernel/Sirius.java (hunk 8 above)
Comment (qw3ry) [reply src/main/java/sirius/kernel/Sirius.java:73]: Not sure if this is necessarily a bad thing. Compared to prod there is much less traffic, so the logs shouldnt be too much to handle, even with debug logging

File: src/main/java/sirius/kernel/Setup.java (hunk 7 above)
Comment (marvinwoop) [reply src/main/java/sirius/kernel/Setup.java]: 👍  Thanks, corrected it.

**Review by andyHa — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Pull request merged**